src/main.cpp

```diff
@@ -6008,10 +6008,13 @@ bool static ProcessMessage(CNode* pfrom, string strCommand, CDataStream& vRecv,
                             pindexLast->GetBlockHash().ToString(), pindexLast->nHeight);
                 }
                 if (vGetData.size() > 0) {
-                    if (nodestate->fSupportsDesiredCmpctVersion && vGetData.size() == 1 && mapBlocksInFlight.size() == 1 && pindexLast->pprev->IsValid(BLOCK_VALID_CHAIN)) {
+                    if (nodestate->fSupportsDesiredCmpctVersion && vGetData.size() == 1 && mapBlocksInFlight.size() == 1 && pindexLast->pprev->IsValid(BLOCK_VALID_CHAIN) && mempool.size() > 1) {
                         // We seem to be rather well-synced, so it appears pfrom was the first to provide us
-                        // with this block! Let's get them to announce using compact blocks in the future.
-                        MaybeSetPeerAsAnnouncingHeaderAndIDs(nodestate, pfrom, connman);
+                        // with this block! Let's get them to announce using compact blocks in the future, unless
+                        // either of us are blocks-only and would prefer to save bandwidth.
+                        if (pfrom->fRelayTxes && !GetBoolArg("-blocksonly", DEFAULT_BLOCKSONLY)) {
```

Comment:
I'm not sure why we wouldn't ask this peer for compact blocks just because they are not relaying transactions.  Our mempool might be perfectly good for block reconstruction, even if the peer who is relaying blocks to us fastest is blocksonly, no? 

I know that in the future we're likely to want to optimize the prefilled transactions based eg on what was missing from our own mempools, but I think at the time we do that we might as well start scoring our peers based on, say, how often a roundtrip is required or something.  And if a blocksonly peer happens to be really fast to deliver us blocks then great...